Add chat icon to the bottom right of all pages

harsh-jindal opened the pull request:
Fixes #2424 
Changes: 
Things fixed till now:
- [x] bottom right placement of chat bubble in laptop view.
- [x] limited to move only in the viewport
- [x] full screen view now opens up, a modal with chat
- [x] increased the width and height made it similar to https://www.intercom.com/live-chat
- [x] Added padding to the full screen and close icons.
- [x] Fixed  "mic icon overlapping white and grey area"
- [x] Fixed ` After dragging the chat bubble, unable to write anything` 
- [x]  Fixed chat bubble placement in mobile and tab view.

Things to be fixed/added:
-  Have search bar next to the full screen and close icon (Shall be implemented in next PR)
Demo Link: https://pr-2465-fossasia-susi-web-chat.surge.sh

Screenshots of the change: [Add screenshots depicting the changes.]
![Screenshot 2019-07-09 at 9 08 05 PM](https://user-images.githubusercontent.com/31013104/60902730-bd358300-a28d-11e9-8a07-54f413ae4c10.png)

harsh-jindal commented:
@mariobehling @akshatnitd @fragm3 This PR is still WIP but can you have look and see whether the changes are being made in the expected files. I have made changes in MessageSection.react.js
Currently only two things are left:
1. Bring it to bottom right corner when the page loads.
2. Be able to move the chat bubble only in window.

mariobehling commented:
Also please see more details on issue page, e.g. include bar to minimize etc. like Google Hangout.

mariobehling commented:
1. Chat bubble hidden behind side bar.
![Screenshot from 2019-07-09 23-14-57](https://user-images.githubusercontent.com/1583873/60923615-94b68480-a29f-11e9-9aa6-2de441689ae6.png)

2. Expanding icon links to /chat instead of expanding window. It also does not provide a way to pop it in again.

harsh-jindal commented:
Fixed `mobile view` and `Don't let the chat box go out of the screen viewport.`
@samagragupta Close your console and then load the page. There is no issue with placement then.

mariobehling commented:
Please position the icon it in the same way as https://www.intercom.com/live-chat

![Screenshot from 2019-07-10 07-37-32](https://user-images.githubusercontent.com/1583873/60943304-d538f100-a2e5-11e9-9feb-ed318a0ad47f.png)

mariobehling commented:
* Also increase width and height of the chat box
* Increase space of X to right side edge 
![Screenshot from 2019-07-10 08-02-20](https://user-images.githubusercontent.com/1583873/60944433-18e12a00-a2e9-11e9-81ab-7b9755bd89cc.png)
* Show scroll bar by default

mariobehling commented:
1. Chat type box does have same edge to each side and mic icon overlapping white and gray area.
2. Voice input does not work. The implementation does not let the user input commands per voice.

Overall this issue is implemented very superficially. There are many website that have a chat icon and it would be a matter of seconds to find them and check how others implemented this. Features are tested superficially as well. You drop the responsibility for checking your PRs on other co-developers and contributors without checking things in depth yourself. Unfortunately, I have to say, this is a lazy and irresponsible behavior towards your co-developers.

harsh-jindal commented:
@mariobehling This is still the version one of the issue. Working on it. Couple of UI enhancements are there, working on it. My main aim in this was to bring it to bottom right corner and and make it draggable. Other websites have chatbots I have seen it, but they don't allow you to move the chat bubble around, that is why it took me time to figure it out. I know the issues you have pointed out I am working on it, will be fixed by night. I requested for review to ask about the positioning and other UI related stuff.

harsh-jindal commented:
@mariobehling The voice input works in the surge link. The text input only works when we toggle the chat bubble, fixing it. Please let me know if it doesn't work for you.

mariobehling commented:
@harsh-jindal We only have a small team and people need to focus on their issues as well. If you have a specific question you can always ask on the chat channel, but if a PR is not ready for review please use the Draft mode or add [WIP] to it. Obvious points you see yourself should be solved before you request a review. Thank you.

harsh-jindal commented:
@mariobehling Sure, will follow that.

mariobehling commented:
* Chat icon has different position on different pages
* Please take out the "Chat" item from the top menu in this PR as well.

harsh-jindal commented:
Things fixed till now:
- [x] bottom right placement of chat bubble in laptop view.
- [x] limited to move only in the viewport
- [x] full screen view now opens up, a modal with chat
- [x] increased the width and height made it similar to https://www.intercom.com/live-chat
- [x] Added padding to the full screen and close icons.
- [x] Fixed  "mic icon overlapping white and grey area"

Things to be fixed/added:
-  After dragging the chat bubble, unable to write anything (Working on this, unable to find solution for this unexpected behaviour. Will try to fix it asap)
-  Have search bar next to the full screen and close icon (Shall be implemented in next PR)
-  Fix chat bubble placement in mobile and tab view. Currently it appears on the bottom right but the placement is not as perfect as laptop view. (Shall be fixed in next PR)

fragm3 commented:
@harsh-jindal Sometimes, unable to type in anything in chat messenger 

@mariobehling We should use current `/chat`full mode, it's better than this PR full mode. We have a lot of space estate in web apps

harsh-jindal commented:
@fragm3 ` Sometimes, unable to type in anything in chat messenger` working on this. It works well in the full screen view but not in the chat view.

harsh-jindal commented:
@fragm3 It is fixed now, please have a look.

mariobehling commented:
@fragm3 
> Why do we have to remove chat from the top link, we can have a discussion on this.

We can have a chat icon in the top bar as well similar to Facebook. We can also do this later.
 
> Points which I feel are not correct:
> 
> 1. I don't think we need it to be draggable. After I have used it, I feel it's really weird to have it
> 2. If I type in something and want to use my mouse to select/edit/delete text, the messenger window shifts. Unable to edit the text using my mouse. I think removing draggable would fix this as well.
> 3. Make the toggle icon like this:

@harsh-jindal Yes, please take care of those three points. Thanks!

fragm3 commented:
@mariobehling Yes, Facebook has a chat icon. We can remove the Chat Link after further discussion as well. Thanks

harsh-jindal commented:
@mariobehling @akshatnitd @fragm3 Fixed, Please review
Note :  the code for navigation bar is commented out, will be removed in next PR after implementing the search.

mariobehling commented:
@harsh-jindal Please resolve conflicts.

mariobehling commented:
![Screenshot from 2019-07-11 09-09-43](https://user-images.githubusercontent.com/1583873/61029649-d0467100-a3bb-11e9-859f-f24b734e8212.png)

Please check that the icon is placed in the same way as on other chat services like intercom.com. Also compare each page, e.g. it moves slightly on the settings page: https://pr-2465-fossasia-susi-web-chat.surge.sh/settings

Intercom
![Screenshot from 2019-07-11 09-13-23](https://user-images.githubusercontent.com/1583873/61029800-26b3af80-a3bc-11e9-8e2d-623cc791b613.png)

mariobehling commented:
Opened chat larger than height of page.

![Screenshot from 2019-07-11 09-17-39](https://user-images.githubusercontent.com/1583873/61030057-d25cff80-a3bc-11e9-84a6-aac1df610a29.png)

harsh-jindal commented:
@mariobehling Fixed , please review. The chat bubble position is fixed for all views.

mariobehling commented:
Thanks. Please make it at the same position as intercom.com. Please compare: https://www.intercom.com/live-chat

harsh-jindal commented:
@mariobehling Can you please attach a screenshot of how it looks for you ?

mariobehling commented:
Please check how intercom does it. The icon stays at the same place at different resolutions. SUSI.AI is a bit too close to the right side bar. Can you use the console inspector and check how they do it?

See here in different resolutions:
![Screenshot from 2019-07-11 12-56-53](https://user-images.githubusercontent.com/1583873/61045625-81a8cf00-a3db-11e9-9efd-0c3998b91788.png)
![Screenshot from 2019-07-11 12-55-59](https://user-images.githubusercontent.com/1583873/61045626-81a8cf00-a3db-11e9-800b-51eb50ee996f.png)
![Screenshot from 2019-07-11 12-55-27](https://user-images.githubusercontent.com/1583873/61045627-81a8cf00-a3db-11e9-86ad-010ad1d8baf5.png)

harsh-jindal commented:
@mariobehling Could also tell the browser you are using ? This looks browser specific to me because for me it sticks next to the scrollbar. Unlike in your screenshots.
`I have marked the area with black line`
![Screenshot 2019-07-11 at 4 56 53 PM](https://user-images.githubusercontent.com/31013104/61047490-40291b80-a3fd-11e9-8d6f-eccc6d1a2275.png)

mariobehling commented:
Issue appears on all three browsers I have Chromium, Chrome, Firefox on Linux Mint 19.

mariobehling commented:
Chat window does not show last chat. User has to scroll down to last chat.
Dark Gray chat bar is too big. Please decrease height.

![Screenshot from 2019-07-11 13-38-36](https://user-images.githubusercontent.com/1583873/61048140-51642f00-a3e1-11e9-9be1-46f22611ad54.png)

mariobehling commented:
On expanded screen the gray vertical sidebar on the left and right of the chat window does not show up.

![Screenshot from 2019-07-11 13-41-36](https://user-images.githubusercontent.com/1583873/61048285-bb7cd400-a3e1-11e9-8ef0-8ee575068de9.png)

mariobehling commented:
Sideline should be like in previous full screen mode.

![Screenshot from 2019-07-11 13-44-18](https://user-images.githubusercontent.com/1583873/61048421-0b5b9b00-a3e2-11e9-952b-54471faae956.png)

akshatnitd commented on package.json:
```diff
@@ -47,6 +47,7 @@
     "react-custom-scrollbars": "4.2.1",
     "react-diff-viewer": "^1.1.0",
     "react-dom": "^16.4.0",
+    "react-draggable": "^3.3.0",
```
Remove it.

harsh-jindal replied:
@akshatnitd I thought we could keep this. In future if we want to make it draggable we will need to just add one line. Shall I remove it still?

fragm3 replied:
Remove it

harsh-jindal commented:
@mariobehling Fixed the chat bubble position.
Regarding the chat not scrolling to the bottom it exits in the current `/chat` route too. Will take that up in my next PR.
@akshatnitd @fragm3 Please review.
Travis is failing due to prettier error in `reducers/create.js` which is strange because i have not made any changes in that.

mariobehling commented:
@harsh-jindal Ok, let's follow up in more issues. Please fix travis.

fragm3 commented on src/components/ChatApp/MessageSection/MessageSection.react.js:
```diff
@@ -528,82 +685,135 @@ class MessageSection extends Component {
+    return (
+      <div>
+        {chatBubble ? (
+          <Draggable
```
Clean this part, Remove Draggable

fragm3 commented:
@harsh-jindal You will have to remove it. It takes a lot of time to write clean code and, I do not want any unwanted code being pushed. It would be hard for a newbie to understand, as well for us.

fragm3 commented on src/components/ChatApp/MessageSection/MessageSection.react.js:
```diff
@@ -114,6 +125,101 @@ const urlPropsQueryConfig = {
+const ChatBubbleContainer = styled.div`
+  position: fixed;
+  // top: ${props => props.height - 90 + 'px'};
+  // right: 85px;
```
Please look into this

harsh-jindal commented:
@fragm3 Will remove it now. That is why I have requested for review before, because i spent my evening fixing it to bottom right and I have to again spend time on it. Also first we want it to be draggable and then not. I am not complaining but if we have atleast a rough idea of what we want clearly would be better because it is very difficult to keep changing one PR because priorities keep changing everyday.

fragm3 commented:
@harsh-jindal I agree, I prefer to look at the code once I see changes fine in surge link. Things are uncertain when it comes to working alone, without any goal of how things proceed. Happens to me as well, but these are core components and it's our responsibility to keep the code clean.

harsh-jindal commented:
@fragm3 Fixed. I have commented from where the props are passed in MessageSection. Please review.

fragm3 commented on src/components/ChatApp/ChatApp.css:
```diff
@@ -4,14 +4,6 @@
   text-align: left;
 }
 
-.scroll::-webkit-scrollbar {
```
Do not remove this

fragm3 commented on src/components/ChatApp/ChatApp.css:
```diff
@@ -4,14 +4,6 @@
-  display: none;
- }
- 
-.scroll {
```
Same here

fragm3 commented on src/redux/reducers/ui.js:
```diff
@@ -24,12 +26,18 @@ export default handleActions(
           isModalOpen: true,
           ...payload,
         },
+        chatBubble: false,
```
Use a string to save multiple states, 'minimized', 'bubble', 'full'

fragm3 commented:
I am trying to go to `/chat` and it's not opening. Be careful, do not break existing functionality

mariobehling commented:
> I am trying to go to `/chat` and it's not opening. Be careful, do not break existing functionality

I think no need anymore as we have the expanding chat.

harsh-jindal commented:
@fragm3 Fixed. The chat shows up on `/chat` route.

harsh-jindal commented:
@mariobehling Fixed travis

akshatnitd commented on src/components/ChatApp/MessageSection/MessageSection.react.js:
```diff
@@ -478,6 +610,25 @@ class MessageSection extends Component {
   }
 
+  openFullScreen = () => {
+    this.props.actions.openModal({ modalType: 'chatBubble' });
```
You can pass fullScreen prop for ChatApp here, no need to make a separate `ChatAppDialog` component.

akshatnitd commented on src/components/Dialog/DialogSection.react.js:
```diff
@@ -79,7 +86,7 @@ const DialogSection = props => {
   };
 
-  const { size, Component, fullScreen = false } = getDialog();
+  const { size, Component, fullScreen = false, padding } = getDialog();
```
Add a default value of padding incase no padding is passed.

harsh-jindal replied:
@akshatnitd If no padding is passed is being handled in shared Container. The reason I have not passed a default prop is because default value will change according to the screen width. 
what i can do is `padding = isMobileView() ? '' : ''`. Shall I give a default value like this ?

akshatnitd commented on src/redux/reducers/ui.js:
```diff
@@ -24,12 +25,14 @@ export default handleActions(
       return {
         ...state,
         modalProps: defaultState.modalProps,
+        chatBubble: payload ? payload.chatBubble : 'bubble',
```
Why are you mixing modal reducers and chat bubble?

harsh-jindal replied:
@akshatnitd This is done because in full screen we don't want the chat bubble. Also chat bubble shall not be present whenever a dialog opens up. So its is placed inside the ui reducer.

akshatnitd replied:
For that the zIndex of Modal should be higher than Chat bubble or dispatch another action to toggle bubble state.

harsh-jindal replied:
@akshatnitd Made the changes.

mariobehling commented:
Please also fix this:
1. Go to the detail page of a skill
2. Click on the example in the speech bubble
-> It should open in the chat box
Currently it opens another page.

Compare
https://susi.ai/chat?testExample=play%20Shakira%20on%20YouTube

![Screenshot from 2019-07-12 08-48-07](https://user-images.githubusercontent.com/1583873/61108196-f898a300-a481-11e9-90fc-1765b365f148.png)

harsh-jindal commented:
@akshatnitd @fragm3 Fixed. Please review.


> Please also fix this:
> 
> 1. Go to the detail page of a skill
> 2. Click on the example in the speech bubble
>    -> It should open in the chat box
>    Currently it opens another page.
> 
> Compare
> https://pr-2465-fossasia-susi-web-chat.surge.sh/chat?testExample=Suggest%20an%20anime
> 
> ![Screenshot from 2019-07-12 08-48-07](https://user-images.githubusercontent.com/1583873/61108196-f898a300-a481-11e9-90fc-1765b365f148.png)

@mariobehling I'll take this and search in my following PR.

harsh-jindal commented:
Search bar and Click on the example in the speech bubble should open in the chat box will be solved in  #2523

fragm3 commented on src/components/ChatApp/MessageComposer.react.js:
```diff
@@ -476,6 +487,8 @@ class MessageComposer extends Component {
             background={textarea}
             color={textcolor}
+            onClick={this.handleClick}
+            key={key}
```
Is this key required? If yes, Please explain why it is being used.

harsh-jindal replied:
@fragm Yes when we click outside of message composer the text area becomes out of focus and clicking again it does not focus again. So I have to update key state on clicking inside the text area so that it is focused again.

mariobehling commented:
@akshatnitd If you find any issues in this PR, please add comments in follow up PRs. I consider this Work in Progress. Merging this, so we can have this blocker for other issues out of the way.
